server/service/src/sync/translations/special/name_merge.rs

```diff
@@ -158,5 +210,87 @@ mod tests {
 
         name_links.sort_by_key(|i| i.id.to_owned());
         assert_eq!(name_links, expected_name_links);
+
+        // When 2 names are merged, we clean up name_store_joins to ensure there only remains 1 NSJ for each store that
+        // had the "deleted" name visible.
+        // e.g. for nameD and nameK where nameD is merged into nameK:
+        // storeA has just nameD visible, so we leave the NSJ and rely on the name_link being updated correctly in the merge process
+        // storeB has both nameD and nameK visible. After the merge, the store has effectively 2 identical name_store_joins pointing to nameK. Thus we delete the name_store_join pointing to nameD.
+        // This prevents names (nameK in this case) showing twice in lists after a merge.
+        // What's more, a store shouldn't become visible to itself! e.g. if storeA above is actually nameK.
+        let (_, connection, _, _) = setup_all(
+            "test_name_merge_message_translation_removes_duplicate_name_store_joins",
+            MockDataInserts::none()
+                .units()
+                .names()
+                .stores()
+                .name_store_joins(),
+        )
+        .await;
+        let mut logger = SyncLogger::start(&connection).unwrap();
+        let name_store_join_repo = NameStoreJoinRepository::new(&connection);
+
+        let count_name_store_join = |id: &str| -> usize {
+            name_store_join_repo
+                .query(Some(
+                    NameStoreJoinFilter::new().name_id(EqualFilter::equal_to(&id.to_string())),
+                ))
+                .unwrap()
+                .len()
+        };
+
+        // Ensure the test data is what was expected as when written
+        assert_eq!(count_name_store_join(&"name_a"), 3);
+        assert_eq!(count_name_store_join(&"name2"), 1);
+        assert_eq!(count_name_store_join(&"name3"), 2);
+        assert_eq!(count_name_store_join(&"name_store_a"), 1);
```

Comment:
Nice, but you could include a check that they all point to store_a :)

Reply:
I'm not sure what you mean? name_store_a is store_a? You mean `name_store_join.store_id == "store_a"`? In that last case, there are actually couple others in the data set. I should test that on the outcome too I am being a little lazy...

```
SELECT nsj.store_id, nsj.id, nl.id, nl.name_id, name.name, name.id   FROM name_store_join nsj 
JOIN name_link nl ON nl.id = nsj.name_link_id
JOIN name ON name.id = nl.name_id
ORDER BY name.id
```

| store_id                     | id                                 | id                            | name_id                       | name                     | id                            |
| ---------------------------- | ---------------------------------- | ----------------------------- | ----------------------------- | ------------------------ | ----------------------------- |
| mock_test_name_query_store_2 | mock_name_1_join_id                | name1                         | name1                         | name_1                   | name1                         |
| store_c                      | name_linked_to_store_a_id          | name_linked_to_store_a_id     | name_linked_to_store_a_id     | Name linked to store a   | name_linked_to_store_a_id     |
| store_a                      | name_linked_to_store_id            | name_linked_to_store_id       | name_linked_to_store_id       | Name linked to store     | name_linked_to_store_id       |
| store_c                      | name_not_linked_to_store_a_id      | name_not_linked_to_store_a_id | name_not_linked_to_store_a_id | Name not linked to store | name_not_linked_to_store_a_id |
| store_a                      | name_not_linked_to_store_id        | name_not_linked_to_store_id   | name_not_linked_to_store_id   | Name not linked to store | name_not_linked_to_store_id   |
| store_a                      | mock_name_store_join_d             | name_a                        | name_store_a                  | Store A                  | name_store_a                  |
| store_c                      | mock_name_store_join_e             | name_a                        | name_store_a                  | Store A                  | name_store_a                  |
| program_master_list_store    | mock_name_store_master_list_join_a | name_store_a                  | name_store_a                  | Store A                  | name_store_a                  |
| mock_test_name_query_store_1 | mock_name_2_join_id                | name2                         | name_store_a                  | Store A                  | name_store_a                  |
| mock_test_name_query_store_2 | mock_name_3_join2_id               | name3                         | name_store_a                  | Store A                  | name_store_a                  |
| store_a                      | name_store_join_b                  | name_store_b                  | name_store_b                  | Store B                  | name_store_b                  |
| program_master_list_store    | mock_name_store_master_list_join_b | name_store_b                  | name_store_b                  | Store B                  | name_store_b                  |
| store_a                      | name_store_join_c                  | name_store_c                  | name_store_c                  | Store C                  | name_store_c                  |
| mock_test_name_query_store_1 | name_a_umlaut_join_id              | name_äÄ_umlaut                | name_äÄ_umlaut                | a*umlaut*äÄ_name         | name_äÄ_umlaut                |
| store_a                      | mock_patient_store_join_b          | patient2                      | patient2                      | patient2                 | patient2                      |
| store_a                      | mock_patient_store_join            | testId                        | testId                        | testId                   | testId                        |


One thing I've realised is to be precise it'd be nice to be able to just test `try_translate_pull_merge` outputs the right `IntegrationRecords`... should I do that today? Haha

Reply:
Oh yes sorry, they not all point to the same store (which wouldn't make any sense). It's just a bit hard to see that this data matches the test cases described above.

What we did before is to just inject new test specific data that is tailored to the current test case.